src/main/java/org/gradle/java/JigsawPlugin.java

```diff
@@ -0,0 +1,231 @@
+
+    private static final String JAVA_PLUGIN = "org.gradle.java";
+
+    private static final String JAVA_LIBRARY_PLUGIN = "org.gradle.java-library";
+
+    private static final String APPLICATION_PLUGIN = "application";
+
+    private static final String EXTENSION_NAME = "architecture";
+
+    private static final String LIBS_PLACEHOLDER = "APP_HOME_LIBS_PLACEHOLDER";
+
+    @Override
+    public void apply(Project project) {
+        LOGGER.debug("Applying JigsawPlugin to " + project.getName());
+        if (!project.getPlugins().hasPlugin(JAVA_PLUGIN) && !project.getPlugins().hasPlugin(JAVA_LIBRARY_PLUGIN)) {
+            project.getPluginManager().apply(JavaPlugin.class);
+        }
+        project.getExtensions().create(EXTENSION_NAME, Architecture.class);
+
+        configureJavaTasks(project);
+    }
+
+    private void configureJavaTasks(final Project project) {
+        project.afterEvaluate(new Action<Project>() {
+            @Override
+            public void execute(final Project project) {
+                configureCompileJavaTask(project);
+                configureCompileTestJavaTask(project);
+                configureJarTask(project);
+                configureTestTask(project);
+                project.getPluginManager().withPlugin(APPLICATION_PLUGIN, new Action<AppliedPlugin>() {
+                    @Override
+                    public void execute(AppliedPlugin appliedPlugin) {
+                        configureRunTask(project);
+                        configureStartScriptsTask(project);
+                    }
+                });
+            }
+        });
+    }
+
+    private void configureCompileJavaTask(final Project project) {
+        final JavaCompile compileJava = (JavaCompile) project.getTasks().findByName(JavaPlugin.COMPILE_JAVA_TASK_NAME);
+        final Architecture module = (Architecture) project.getExtensions().getByName(EXTENSION_NAME);
+        final String moduleDir =
+                compileJava.getDestinationDir().toString() + System.getProperty("path.separator") + module.getModuleName();
+        compileJava.getInputs().property("moduleName", module.getModuleName());
+        compileJava.doFirst(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                List<String> args = new ArrayList<>();
+                args.add("--module-path");
+                args.add(compileJava.getClasspath().getAsPath());
+                args.add("-d");
+                args.add(moduleDir);
+                compileJava.getOptions().setCompilerArgs(args);
+                compileJava.setClasspath(project.files());
+            }
+        });
+        compileJava.doLast(new Action<Task>() {
+            @Override
+            public void execute(Task task) {
+                project.copy(new Action<CopySpec>() {
+                    @Override
+                    public void execute(CopySpec copySpec) {
+                        copySpec.from(moduleDir);
```

Comment:
So, this is part of the "quick-and-dirty" aspect of this hackery. I guess I should explain it in the comments in the code as well, but I'll write it here first to see if it makes sense and also to see if you have a more clever solution.

Here are the design constraints which lead to this hack.
1. Don't require major changes to Gradle.
2. Allow piece-meal conversion from the bottom up. That is, just one project in a multi-project build can become modular without the rest.
3. Allow for other already modular projects to consume the converted project.

So, a modular jar file is also a regular jar file. So, it can be used on the `compileClasspath` or the `runtimeClasspath` in a non-modular project. Likewise, it could be used on the `compileClasspath` or `runtimeClasspath` of a modular project which would then copy that classpath to the `--module-path` argument to `java` or `javac`.

*But*, when we have project-to-project dependencies, we don't put the jar file on the classpath, we depend directly on the `destinationDir` of the `compileJava` task. Unfortunately, directories on the `--module-path` are interpreted differently from directories passed to the `-classpath` option. Every class in a module on the `--module-path` must be rooted one level below the directory listed on the module path. So, if the compiler is looking for class `org.gradle.actors.Actor` on a classpath with directory `classes`, it will look for `classes/org/gradle/actors/Actor.class` but on the module path, that same class would be expected to live in `classes/someDir/org/gradle/actors/Actor.class` where the corresponding `module-info.class` is in `classes/someDir`. Now, `someDir` doesn't have to be named `moduleName` but it doesn't hurt.

So, if we want to be able to use the `destinationDir` for the task on the classpath when consumed by projects that are not yet modularized AND on the module path when consumed by projects that are already modularized then we have to have the classes in both places.

You talk about needing the `classes` dir to be `clean`. But, does it really need to be? Nobody should be importing `import moduleName.org.gradle.actors.Actor;` so, what's the harm?

Obviously, for the "real", non-hacky solution, we'll want to do something more sophisticated. Maybe, use our own `JavaFileManager` to tell the compiler where to find what it's looking for, but trying to just use the public API available to build script and plugin authors, this is what I was able to come up with. I'm not saying it's ideal. But, it gets the job done.

All that being said, if I'm missing an easier approach (likely), I'm really happy to learn of it.